typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+  const newArray: string[] = [];
+  for (let i = 0; i < arr.length; i++) {
+    if (fn(arr[i])) {
+      newArray.push(arr[i]);
+    }
+  }
+  return newArray; // or just simply return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
+  return (s: number) => {
+    return s.toString() === x;
+  };
+};
+
+f("5")(5) // true;
+```
+
+As you may see in the last example, we can call the function `f` and return a function from it, applying in for the new argument.
+So, yes, we can implement some kind of currying through this.
+
+You can split type signature of the function from its implementation.
+Unfortunately, unnamed parameters are not supported, but you can set them as `_`.
+
+```typescript
+type addT = (_: number) => (_: number) => number;
+const add: addT = (l) => (r) => l + r;
+
+type foldArrT = (f: (_: number) => (_: number) => number) => (_: number) => (_: number[]) => number;
+const foldArr: foldArrT = (f) => (z) => (arr) => {
+  let m = z;
+  arr.forEach((elem) => {
+    m = f(m)(elem);
+  });
+  return m;
+};
+
+foldArr(add)(0)([1, 2, 3, 4, 5]) // 15;
+```
+
+Unfortunately, the syntax is weird since you need to always use named parameters.
+But it is still a possible functionality, and you can use it in your code quite simply.
+
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad-hoc one and Row one.
+Let's start from the first one.
+
+#### Parametric polymorphism
+
+Parametric polymorphism allows us to write abstract function or data which will not depend on its type.
+In TypeScript, it is called generics.
+
+```typescript
+const arrLength: <T>(_: T[]) => number = (arr) => arr.length;
+
+arrLength([1, 2, 3]); // 3
+arrLength<string>(["1", "2"]); // 2
+arrLength<number>(["1", "2", "3"]); // Type 'string' is not assignable to type 'number'.
+
+type NonEmpty<T> = {
+  head: T;
+  tail: T[];
+};
+```
+
+#### Ad-hoc polymorphism
+
+Ad hoc polymorphism allows you to implement abstract functions, which logic will be different with different types.
+
+
+Let's define interface `Eq`
+
+```typescript
+
+interface Eq<T> {
+  equal: (f: T, s: T) => boolean;
+}
+```
+
+And now two implementations of it: `IntEq` for numbers and `IntArrEq` for arrays of numbers.
+The first one will simply compare two numbers.
+And the second one will return equality if the first array contain all elements from the second array.
+
+```typescript
+class IntEq implements Eq<number> {
+  equal(f: number, s: number) {
+    return f === s;
+  }
+}
+
+class IntArrEq implements Eq<number[]> {
+  equal(f: number[], s: number[]) {
+    return f.filter(item => s.indexOf(item) < 0).length === 0;
+  }
+}
+```
+
+Now let's implement `lookup` function which for key value pairs array and specific key will return the value associated with this key or `undefined`.
+We can write it polymorph passing comparator class which implements `Eq` for the key type.
```

Comment:
I don't grasp this sentence. 

"We can make it polymorphic by passing a comparator class that implements `Eq` for the key type"?